snuba/migrations/snuba_migrations/events/0003_errors.py

```diff
@@ -1,100 +1,99 @@
 from typing import Sequence
 
 from snuba.clickhouse.columns import (
+    UUID,
     Array,
     Column,
     DateTime,
     FixedString,
     IPv4,
     IPv6,
     Nested,
-    Nullable,
     String,
     UInt,
-    UUID,
 )
 from snuba.clusters.storage_sets import StorageSetKey
 from snuba.migrations import migration, operations, table_engines
-from snuba.migrations.columns import (
-    LowCardinality,
-    Materialized,
-    WithCodecs,
-    WithDefault,
-)
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+from snuba.migrations.columns import lowcardinality, nullable
 
 columns = [
     Column("org_id", UInt(64)),
     Column("project_id", UInt(64)),
     Column("timestamp", DateTime()),
-    Column("event_id", WithCodecs(UUID(), ["NONE"])),
+    Column("event_id", (UUID(Modifiers(codecs=["NONE"])))),
     Column(
         "event_hash",
-        WithCodecs(
-            Materialized(UInt(64), "cityHash64(toString(event_id))",), ["NONE"],
+        UInt(
+            64,
+            Modifiers(materialized="cityHash64(toString(event_id))", codecs=["NONE"]),
         ),
     ),
-    Column("platform", LowCardinality(String())),
-    Column("environment", LowCardinality(Nullable(String()))),
-    Column("release", LowCardinality(Nullable(String()))),
-    Column("dist", LowCardinality(Nullable(String()))),
-    Column("ip_address_v4", Nullable(IPv4())),
-    Column("ip_address_v6", Nullable(IPv6())),
-    Column("user", WithDefault(String(), "''")),
-    Column("user_hash", Materialized(UInt(64), "cityHash64(user)"),),
-    Column("user_id", Nullable(String())),
-    Column("user_name", Nullable(String())),
-    Column("user_email", Nullable(String())),
-    Column("sdk_name", LowCardinality(Nullable(String()))),
-    Column("sdk_version", LowCardinality(Nullable(String()))),
+    Column("platform", String(lowcardinality())),
+    Column("environment", String(Modifiers(nullable=True, low_cardinality=True))),
+    Column("release", String(Modifiers(nullable=True, low_cardinality=True))),
+    Column("dist", String(Modifiers(nullable=True, low_cardinality=True))),
+    Column("ip_address_v4", IPv4(nullable())),
+    Column("ip_address_v6", IPv6(nullable())),
+    Column("user", (String(Modifiers(default="''")))),
+    Column("user_hash", UInt(64, Modifiers(materialized="cityHash64(user)"))),
+    Column("user_id", String(nullable())),
```

Comment:
I am not a huge fan of having many different conventions for specifying modifiers here. I see that `nullable` is provided to keep things more concise, however I'd personally prefer something a bit more consistent.

If the goal is to make things really concise and make it as easy as possible to for anyone to write migrations, what about going with something like:
`String(nullable=True, low_cardinality=True)` 

where `String` is a migrations specific class (possibly a subclass of the schema's String)

Reply:
I think the `String(nullable=True, low_cardinality=True)` is viable, though wouldn't it still be yet another convention? `String(Modifiers(nullable=True, low_cardinality=True))` would still be accepted.

In that sense I think I would prefer removing `nullable()` and `lowcardinality()` entirely and replace them with 
`String(Modifiers(nullable=True, low_cardinality=True))`

What doo you think ?

Reply:
Well, I mean it's a different convention but it would be would be the only convention we'd use everywhere. Not just for string, but we'd have the equivalent for every other column type and modifier combination, e.g. we'd also have `UInt(8, nullable=True, codec=["NONE"])` and friends. We've remove `nullable()` / `lowcardinality()`. We'd also never import MigrationModifiers in the migration. `MigrationModifiers` would instead get created by the ColumnType class. We'd also use a similar syntax for schemas and SchemaModifiers.

I agree that we can remove `nullable()` and `lowcardinality()` entirely whichever way we go, but I do think there's value in making the schemas as simple and easy to write as possible.

Reply:
> we'd also have UInt(8, nullable=True, codec=["NONE"]) and friends

I think this would become a problem. (See the PR on top of this one: https://github.com/getsentry/snuba/pull/1443). Doing something like that would require us to duplicate all column types. The migration columns could not extend the schema columns (SchemaModifiers and MigrationModifiers are not one the subclass of the other intentionally). 

In order to reuse types across schemas and migrations The constructor either has to be the same or the migration one must be an extension of the schema one. That way I think what we loose in higher verbosity is better than fully duplicating the types.

One alternative is to let the classes the way they are with this PR and add helper methods to instantiate them instead of helper methods to provide modifiers:
```
class String:
   ... whatever it is now in this PR ...

# In clickhouse.columns
def string(nullable: bool, readonly: bool) -> String:
    return String(Modifiers(nullable=....))

# in migrations.py
def string(nullable: bool, lowcardinality: bool, default: string, etc.) -> String:
    return String(MigrationModifiers(.....))
```

Though I would do this addition in a separate PR since this is already mroe than 1000 lines long. Here I would just remove nullable() and readonly().

What about this ?